File: Grid/Objc/GridViewExample/MainGridController.m
//
//  MainGridController.m
//  GridViewExample
//
//  Created by Adam Fish on 10/27/15.
//  Copyright © 2015 Adam Fish. All rights reserved.
//

#import "MainGridController.h"
#import "MainCollectionViewCell.h"

#import <RealmNYTStories/NYTStory.h>
#import <RealmNYTStories/NYTStoryImage.h>
#import <Realm/Realm.h>
#import <Haneke/Haneke.h>
#import <TOWebViewController/TOWebViewController.h>

@interface MainGridController ()

@end

@implementation MainGridController

static NSString * const reuseIdentifier = @"mainCell";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.entityName = @"NYTStory";
    
    self.sortDescriptors = @[[RLMSortDescriptor sortDescriptorWithProperty:@"publishedDate" ascending:NO]];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    MainCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier
                                                                             forIndexPath:indexPath];
    
    // Configure the cell
    NYTStory *story = [self objectAtIndexPath:indexPath];
    
    cell.titleLabel.text = story.title;
    cell.dateLabel.text = [NYTStory stringFromDate:story.publishedDate];
    cell.excerptLabel.text = story.abstract;
    
    // Use Haneke image caching
    [cell.imageView hnk_setImageFromURL:story.storyImage.url];
    
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    [collectionView deselectItemAtIndexPath:indexPath animated:YES];
    
    NYTStory *story = [self objectAtIndexPath:indexPath];
    
    TOWebViewController *webController = [[TOWebViewController alloc] initWithURLString:story.urlString];
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:webController];
    
    [self.navigationController presentViewController:navController animated:YES completion:nil];
}

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    CGFloat height = 250.0;
    
    if ([[UIApplication sharedApplication] statusBarOrientation] == UIInterfaceOrientationPortrait) {
        CGFloat columns = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad ? 3.0 : 2.0;
        
        CGFloat width = CGRectGetWidth(self.view.frame) / columns;
        
        return CGSizeMake(width, height);
    }
    else {
        CGFloat columns = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad ? 4.0 : 3.0;
        
        CGFloat width = CGRectGetWidth(self.view.frame) / columns;
        
        return CGSizeMake(width, height);
    }
}

- (IBAction)didClickRefreshButton:(UIBarButtonItem *)sender
{
    [NYTStory loadLatestStoriesIntoRealm:[RLMRealm defaultRealm]
                              withAPIKey:@"388ce6e70d2a8e825757af7a0a67c397:13:59285541"];
}

@end

File: Grid/Swift/GridViewExample/MainGridController.swift
//
//  MainGridController.swift
//  GridViewExample
//
//  Created by Adam Fish on 10/26/15.
//  Copyright © 2015 Adam Fish. All rights reserved.
//

import UIKit
import RealmSwiftNYTStories
import RealmGridController
import RealmSwift
import Haneke
import TOWebViewController

private let reuseIdentifier = "mainCell"

class MainGridController: RealmGridController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set the Realm object class name that the grid controller will bind to
        self.entityName = "NYTStory"
        
        // Sort the article in reverse chronological order
        self.sortDescriptors = [SortDescriptor(property: "publishedDate", ascending: false)]
    }

    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(reuseIdentifier, forIndexPath: indexPath) as! MainCollectionViewCell
    
        // Configure the cell
        let aStory = self.objectAtIndexPath(NYTStory.self, indexPath: indexPath)
        
        cell.titleLabel.text = aStory?.title
        cell.excerptLabel.text = aStory?.abstract
        
        if let date = aStory?.publishedDate {
            cell.dateLabel.text = NYTStory.stringFromDate(date)
        }
        
        // Lazy load image using Haneke library
        if let imageURL = aStory?.storyImage?.url {
            cell.imageView.hnk_setImageFromURL(imageURL)
        }
    
        return cell
    }
    
    override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        collectionView.deselectItemAtIndexPath(indexPath, animated: true)
        
        let story = self.objectAtIndexPath(NYTStory.self, indexPath: indexPath)
        
        if let urlString = story?.urlString {
            let webController = TOWebViewController(URLString: urlString)
            
            let navController = UINavigationController(rootViewController: webController)
            
            self.navigationController?.presentViewController(navController, animated: true, completion: nil)
        }
    }
    
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        let height: CGFloat = 250.0
        
        if UIApplication.sharedApplication().statusBarOrientation == UIInterfaceOrientation.Portrait {
            let columns: CGFloat = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiom.Pad ? 3.0 : 2.0
            
            let width = CGRectGetWidth(self.view.frame) / columns
            
            return CGSizeMake(width, height)
        }
        else {
            let columns: CGFloat = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiom.Pad ? 4.0 : 3.0
            
            let width = CGRectGetWidth(self.view.frame) / columns
            
            return CGSizeMake(width, height)
        }
    }

    @IBAction func didClickRefreshButton(sender: UIBarButtonItem) {
        NYTStory.loadLatestStories(intoRealm: try! Realm(), withAPIKey: "INSERT_YOUR_API_KEY_HERE");
    }
}
